src/main/java/io/anserini/collection/CommonCrawlWARCCollection.java

```diff
@@ -0,0 +1,746 @@
+        throw new NoSuchElementException();
+      }
+
+      // now read the bytes of the content
+      retContent = new byte[contentLength];
+      int totalWant = contentLength;
+      int totalRead = 0;
+      while (totalRead < contentLength) {
+        try {
+          int numRead = in.read(retContent, totalRead, totalWant);
+          if (numRead < 0) {
+            throw new NoSuchElementException();
+          } else {
+            totalRead += numRead;
+            totalWant = contentLength - totalRead;
+          } // end if (numRead < 0) / else
+        } catch (EOFException eofEx) {
+          // resize to what we have
+          if (totalRead > 0) {
+            byte[] newReturn = new byte[totalRead];
+            System.arraycopy(retContent, 0, newReturn, 0, totalRead);
+            return newReturn;
+          } else {
+            throw new NoSuchElementException();
+          }
+        } // end try/catch (EOFException)
+      } // end while (totalRead < contentLength)
+
+      return retContent;
+    }
+
+    /**
+     * Reads in a WARC record from a data input stream.
+     *
+     * @param in the input stream
+     * @param version WARC version
+     * @return a WARC record (or null if EOF)
+     * @throws IOException if error encountered reading from stream
+     */
+    public static Document readNextWarcRecord(DataInputStream in, String version)
+        throws IOException {
+      StringBuilder recordHeader = new StringBuilder();
+      byte[] recordContent = readNextRecord(in, recordHeader, version);
+
+      // extract out our header information
+      String thisHeaderString = recordHeader.toString();
+      String[] headerLines = thisHeaderString.split(Document.NEWLINE);
+
+      Document retRecord = new Document();
+      for (int i = 0; i < headerLines.length; i++) {
+        String[] pieces = headerLines[i].split(":", 2);
+        if (pieces.length != 2) {
+          retRecord.addHeaderMetadata(pieces[0], "");
+          continue;
+        }
+        String thisKey = pieces[0].trim();
+        String thisValue = pieces[1].trim();
+
+        // check for known keys
+        if (thisKey.equals("WARC-Type")) {
+          retRecord.setWarcRecordType(thisValue);
+        } else if (thisKey.equals("WARC-Date")) {
+          retRecord.setWarcDate(thisValue);
+        } else if (thisKey.equals("WARC-Record-ID")) {
+          retRecord.setWarcUUID(thisValue);
+        } else if (thisKey.equals("Content-Type")) {
+          retRecord.setWarcContentType(thisValue);
+        } else {
+          retRecord.addHeaderMetadata(thisKey, thisValue);
+        }
+      }
+
+      // set the content
+      retRecord.setContent(recordContent);
+
+      return retRecord;
+    }
+  }
+
+
+  /**
+   *
+   * A document from the <a href="https://commoncrawl.org/the-data/get-started/#WARC-Format/">
+   * CommonCrawl WARC collection</a>.
+   *
+   */
+  public static class Document implements CommonCrawlBaseDocument {
```

Comment:
How much of this is duplicated from CW12? Can we use inheritance to reduce copy/pasting of code?